### PR TITLE
Proposal: keep support documentation with code

### PR DESCRIPTION
**Pull request by Floppy**
This is a quick knockup of a proposal to keep support documentation easily discoverable alongside the code it documents. It probably needs more work, but it's open for comment.

The idea is that support docs are inconsistent. Let's make them consistent by providing docs in the code repo, or at the very least providing a link to the docs from the code.

### CODE REVIEW

**Comment by erbridge** [comment]
Missing a file extension @Floppy

**Comment by jacksonj04** [comment]
This sounds entirely sane in principle (although my preference would be for `docs/` over `doc/` - my only worry would be how maintainable the documentation is. Are we in a position where all support agents have write access to all repositories they may want to update the documentation on?

**Comment by erbridge** [comment]
We are currently giving admin access to all developers on all projects, so they should have that power.

**Comment by erbridge** [comment]
> Missing a file extension @Floppy

Also file number is out by 1.

File: rfc-031-keep-support-docs-with-code [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+## Proposal
+
+All source code repositories MUST have a README.md file in the root. This file MUST include a link to support documentation under a `Support` heading.
+
+All repositories MAY contain a `doc` folder. 
+
+If it exists, the `doc` folder MUST include a file named `support.md` which acts as a first port of call for support users.
+
+If the repository has a `doc` folder, the README should link to the `support.md` file within it under this heading.
```
Comment (erbridge) [inline rfc-031-keep-support-docs-with-code:21]:
```suggestion
If the repository has a `doc` folder, the readme SHOULD link to the `support.md` file within it under this heading.
```

File: rfc-031-keep-support-docs-with-code [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+## Problem
+
+Support documentation exists in multiple places currently, and can be confusing. Some projects have `doc` directories, some don't, and some have multiple sources of truth. We should aim to standardise this.
+
+As well as making documentation easier to find, having it in standard locations makes future automation of support documentation easier.
+
+## Proposal
+
+All source code repositories MUST have a README.md file in the root. This file MUST include a link to support documentation under a `Support` heading.
```
Comment (erbridge) [inline rfc-031-keep-support-docs-with-code:15]:
```suggestion
All source code repositories MUST have a `README.md` file in the root. This file MUST include a link to support documentation under a `Support` heading.
```

**Comment by lawrence-forooghian** [comment]
> We are currently giving admin access to all developers on all projects, so they should have that power.

Is that true even for repos that we don't own – e.g. Claim?

File: rfc-031-keep-support-docs-with-code [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+All repositories MAY contain a `doc` folder.
```
Comment (erbridge) [inline rfc-031-keep-support-docs-with-code:17]: I agree with Nick. `docs/` feels more natural. Perhaps we can allow flexibility according to conventions for the project?

```suggestion
All repositories MAY contain a `doc` or `docs` folder.
```

File: rfc-031-keep-support-docs-with-code [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+As well as making documentation easier to find, having it in standard locations makes future automation of support documentation easier.
+
+## Proposal
+
+All source code repositories MUST have a README.md file in the root. This file MUST include a link to support documentation under a `Support` heading.
+
+All repositories MAY contain a `doc` folder. 
+
+If it exists, the `doc` folder MUST include a file named `support.md` which acts as a first port of call for support users.
```
Comment (erbridge) [inline rfc-031-keep-support-docs-with-code:19]:
```suggestion
If it exists, the documentation folder MUST include a file named `support.md` which acts as a first port of call for support users.
```

File: rfc-031-keep-support-docs-with-code (hunk 1 above)
Comment (erbridge) [inline rfc-031-keep-support-docs-with-code:21]:
```suggestion
If the repository has a documentation folder, the README should link to the `support.md` file within it under this heading.
```

File: rfc-031-keep-support-docs-with-code [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+If the repository has a `doc` folder, the README should link to the `support.md` file within it under this heading.
+
+If a repository does not contain a `doc` folder, the support section in the README MUST contain a link to support details, ideally to another related repository that includes `doc` folder.
```
Comment (erbridge) [inline rfc-031-keep-support-docs-with-code:23]:
```suggestion
If a repository does not contain a documentation folder, the support section in the README MUST contain a link to support details, ideally to another related repository that includes a documentation folder.
```

Comment (erbridge) [inline rfc-031-keep-support-docs-with-code:23]:
```suggestion
If a repository does not contain a `doc` folder, the support section in the README MUST contain a link to support details.
```

I'm not sure that adds much.

File: rfc-031-keep-support-docs-with-code [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+`support.md` MUST include details of (or a link to another file which describes) how to:
+
+* run the code in a development environment
+* how to deploy the code
```
Comment (erbridge) [inline rfc-031-keep-support-docs-with-code:28]:
```suggestion
* deploy the code
```

**Comment by erbridge** [comment]
@lawrence-forooghian We should be giving at least write access, or how will they support the project?

File: rfc-031-keep-support-docs-with-code (hunk 5 above)
Comment (CristinaRO) [reply rfc-031-keep-support-docs-with-code:23]: Also, suggest "If a repository does not contain a documentation folder", in line with F's similar suggestions above.

**Comment by CristinaRO** [comment]
> > We are currently giving admin access to all developers on all projects, so they should have that power.
> 
> Is that true even for repos that we don't own – e.g. Claim?

Good question: there are repos where the devs on support only get access on a need to have basis.

There has to be a place that is accessible by dxw devs which documents how to request access, who are the people who can grant it etc. Possibly part of Project Wisdom / Knowledgebase.

**Comment by erbridge** [comment]
Also, what about out of hours support? If things we support are locked down and not visible to people on that rota, they won't be able to get access when they need it.

**Comment by Floppy** [comment]
Thanks everyone for the feedback here, and spotting my obvious mistakes made when rushing :)

Regarding `doc` vs `docs`, I was basing that on the idea that Rails produced a `doc` directory by default, but that turns out to be a complete figment of my imagination, so it can be either. I too prefer `docs` if there's no existing convention to adhere to.

File: rfc-032-keep-support-docs-with-code.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+## Problem
+
+Support documentation exists in multiple places currently, and can be confusing. Some projects have `doc` directories, some don't, and some have multiple sources of truth. We should aim to standardise this.
+
+As well as making documentation easier to find, having it in standard locations makes future automation of support documentation easier.
+
+## Proposal
+
+All source code repositories MUST have a `README.md` file in the root. This file MUST include a link to support documentation under a `Support` heading.
```
Comment (lawrence-forooghian) [inline rfc-032-keep-support-docs-with-code.md:15]: I don't think this applies to all repositories. It's only ones that we have a support contract for.

**Pull request closed without merging**